### PR TITLE
Fix #6773: Don't do image extent adjustment...

### LINKED ISSUE #6773: Shifted image extents in 2.0.0.b3
Comparing images in 1.5.2 and 2.0.0b3 shows a reasonably large change. Originally, I thought the sampling was just different and it was intended, but closer inspection seems to indicate that it might actually be a small shift in the extent.

 I don't have a simple example, but here are two from Cartopy to illustrate what I mean:
global_map_v1.5.2
global_map_v2.0.0b3
If you switch between them, you can see that 2.0.0b3 seems higher, though I think it's actually the right edge that's a bit further out.

A better example is:
imshow_regional_projected_v1.5.2
imshow_regional_projected_v2.0.0b3
It's clear on this one that the image is just a bit wider in 2.0.0b3 vs 1.5.2.

The code for these two plots is [global_map.py](https://github.com/SciTools/cartopy/blob/master/lib/cartopy/examples/global_map.py) and [test_regional_projected](https://github.com/SciTools/cartopy/blob/master/lib/cartopy/tests/mpl/test_images.py#L146).

It looks like one edge is off-by-one or something, but it's also possible that the transforms were just corrected in some way.

### PR DESCRIPTION
**Pull request by mdboom**
...on non-rectilinear axes

### CODE REVIEW

**Comment by QuLogic** [comment]
I don't know if I'm running this PR wrong, but it seems like the Cartopy result is the same here; though somehow Travis thinks differently. Will have to investigate later.

File: lib/matplotlib/image.py [hunk 1]
```diff
@@ -336,11 +338,16 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
...
+        # So that the image is aligned with the edge of the axes, we
+        # want to round up the output width to the next integer.  This
+        # also means scaling the transform just slightly to account
+        # for the extra subpixel.  This is only required for regular
+        # rectangular axes where the edges are pixel-snapped.  For
+        # non-rectangular axes, this refinement isn't required.
+        if (round_to_pixel_border and
+            t.is_affine and
+            # deliberate exact class match
+            type(self.axes) in (maxes.Axes, maxes._subplots.Subplot) and
```
Comment (tacaswell) [inline lib/matplotlib/image.py]: Does this mess up anything with wcaxes?

**Comment by QuLogic** [comment]
OK, I've tested this one out now. It's not quite as expected, but it doesn't go over the edge now. But it seems like the snapping works in reverse now?

Expected:
gshhs_coastlines
vs. this PR:
result-gshhs_coastlines
from [this test](https://github.com/SciTools/cartopy/blob/master/lib/cartopy/tests/mpl/test_features.py#L55).

**Comment by mdboom** [comment]
@QuLogic: What is the axes class used in your example.  This fix uses the class of the Axes to determine whether it has rectilinear edges or not.  I suspect that isn't good enough, and maybe we'll have to do something smarter here.

**Comment by QuLogic** [comment]
`cartopy.mpl.geoaxes.GeoAxesSubplot`

**Comment by mdboom** [comment]
I'm changed the test to be on the Axes.patch, not the Axes class itself.

@QuLogic: How does this work for your use case?

**Comment by QuLogic** [comment]
This does fix the images in #6773, though it breaks Cartopy tests in the opposite way. That is, the image is now "too small", but I think it's actually more correct in that the image is no longer (ever so slightly) outside the expected border.

The example above still produces the white line on the right, but now I'm not sure if it's a bug in `imshow` or in the outline...

**Comment by efiring** [comment]
I'm closing this because #7465 has been merged.

**Pull request closed without merging**